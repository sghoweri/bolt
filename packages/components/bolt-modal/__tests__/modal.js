/* eslint-disable no-await-in-loop */
import {
  render,
  stopServer,
  renderWC,
  html,
  vrtDefaultConfig as vrtConfig,
} from '../../../testing/testing-helpers';

const { readYamlFileSync } = require('@bolt/build-tools/utils/yaml');
const { join } = require('path');
const schema = readYamlFileSync(join(__dirname, '../modal.schema.yml'));
const { width, spacing, theme, scroll } = schema.properties;

const vrtDefaultConfig = Object.assign(vrtConfig, {
  failureThreshold: '0.02',
  customDiffConfig: {
    includeAA: true,
  },
  allowSizeMismatch: true,
});

const timeout = 120000;

// Currently, the only important breakpoints to test are 'small' and 'large'
const viewportSizes = [
  {
    size: 'large',
    width: 1024,
    height: 768,
  },
  {
    size: 'small',
    width: 320,
    height: 568,
  },
];

const modalContent = [
  {
    name: 'Simple usage',
    content: `<bolt-text>Default slot.</bolt-text>`,
  },
  {
    name: 'Long content usage',
    content: `
      <bolt-text subheadline font-size="xlarge">This is very long content.</bolt-text>
      <bolt-image src="/fixtures/1200x2500.jpg" alt="Placeholder"></bolt-image>
    `,
  },
];

describe('<bolt-modal> Component', () => {
  let page;

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  }, timeout);

  beforeAll(async () => {
    page = await global.__BROWSER__.newPage();
    await page.goto('http://127.0.0.1:4444/', {
      timeout: 0,
    });
  }, timeout);

  afterAll(async () => {
    await stopServer();
    await page.close();
  }, timeout);

  test('basic usage', async () => {
    const results = await render('@bolt-components-modal/modal.twig', {
      content: 'This is a modal',
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  width.enum.forEach(async widthChoice => {
    test(`modal width: ${widthChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        width: widthChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  spacing.enum.forEach(async spacingChoice => {
    test(`modal spacing: ${spacingChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        spacing: spacingChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  theme.enum.forEach(async themeChoice => {
    test(`modal theme: ${themeChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        theme: themeChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  scroll.enum.forEach(async scrollChoice => {
    test(`modal scroll: ${scrollChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        scroll: scrollChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  test(
    `<bolt-modal> slots`,
    async () => {
      const { html, ok } = await render('@bolt-components-modal/modal.twig', {
        content: `<bolt-text slot="header">Header slot</bolt-text>Default slot<bolt-text slot="footer">Footer slot</bolt-text>`,
        width: 'regular',
      });
      expect(ok).toBe(true);
      expect(html).toMatchSnapshot();

      await renderWC('bolt-modal', html, page);

      const screenshots = [];

      for (const item of viewportSizes) {
        const { height, width, size } = item;

        screenshots[size] = [];

        await page.setViewport({ height, width });
        await page.evaluate(async () => {
          const modals = Array.from(document.querySelectorAll('bolt-modal'));
          await customElements.whenDefined('bolt-modal');
          return await Promise.all(
            modals.map(elem => {
              if (elem._wasInitiallyRendered) return;
              return new Promise((resolve, reject) => {
                elem.addEventListener('ready', resolve);
                elem.addEventListener('error', reject);
              });
            }),
          ).then(() => {
            modals[0].show();
          });
        });
        await page.waitFor(500);

        screenshots[size].modalOpened = await page.screenshot();
        expect(screenshots[size].modalOpened).toMatchImageSnapshot(
          vrtDefaultConfig,
        );

        await page.evaluate(() => {
          document.querySelector('bolt-modal').hide();
        });
        await page.waitFor(500);
      }
    },
    timeout,
  );

  modalContent.forEach(async contentChoice => {
    test(`${contentChoice.name} <bolt-modal> with Shadow DOM renders`, async () => {
      const { outerHTML } = await renderWC(
        'bolt-modal',
        `
        <bolt-modal uuid="12345" width="regular">
          <bolt-text slot="header">Header slot</bolt-text>
          ${contentChoice.content}
          <bolt-text slot="footer">Footer slot</bolt-text>
        </bolt-modal>
      `,
        page,
      );

      const screenshots = [];
      for (const item of viewportSizes) {
        const { height, width, size } = item;
        screenshots[size] = [];

        await page.setViewport({ height, width });

        await page.evaluate(() => {
          document.querySelector('bolt-modal').show();
        });
        await page.waitFor(500);
        screenshots[size].modalOpened = await page.screenshot();

        expect(screenshots[size].modalOpened).toMatchImageSnapshot(
          vrtDefaultConfig,
        );
        await page.evaluate(() => {
          document.querySelector('bolt-modal').hide();
        });

        await page.waitFor(500);
      }

      const renderedHTML = await html(outerHTML);
      expect(renderedHTML).toMatchSnapshot();
    });
  });

  modalContent.forEach(async contentChoice => {
    test(`${contentChoice.name} <bolt-modal> w/o Shadow DOM renders`, async () => {
      const { outerHTML } = await renderWC(
        'bolt-modal',
        `
        <bolt-modal uuid="12345" width="regular" no-shadow>
          <bolt-text slot="header">Header slot</bolt-text>
          ${contentChoice.content}
          <bolt-text slot="footer">Footer slot</bolt-text>
        </bolt-modal>
      `,
        page,
      );

      const screenshots = [];

      for (const item of viewportSizes) {
        const { height, width, size } = item;

        screenshots[size] = [];

        await page.setViewport({ height, width });
        await page.evaluate(() => {
          document.querySelector('bolt-modal').show();
        });
        await page.waitFor(500);

        screenshots[size].modalOpened = await page.screenshot();
        expect(screenshots[size].modalOpened).toMatchImageSnapshot(
          vrtDefaultConfig,
        );

        await page.evaluate(() => {
          document.querySelector('bolt-modal').hide();
        });
        await page.waitFor(500);
      }

      const renderedHTML = await html(outerHTML);
      expect(renderedHTML).toMatchSnapshot();
    });
  });

  modalContent.forEach(async contentChoice => {
    test(
      `${contentChoice.name} <bolt-modal> at various viewport sizes`,
      async () => {
        const { html, ok } = await render('@bolt-components-modal/modal.twig', {
          content: `<bolt-text slot="header">Header slot</bolt-text>
            ${contentChoice.content}
            <bolt-text slot="footer">Footer slot</bolt-text>`,
          width: 'regular',
        });
        expect(ok).toBe(true);
        expect(html).toMatchSnapshot();

        await renderWC('bolt-modal', html, page);

        const screenshots = [];

        for (const item of viewportSizes) {
          const { height, width, size } = item;

          screenshots[size] = [];

          await page.setViewport({ height, width });
          await page.evaluate(() => {
            document.querySelector('bolt-modal').show();
          });
          await page.waitFor(500);

          screenshots[size].modalOpened = await page.screenshot();
          expect(screenshots[size].modalOpened).toMatchImageSnapshot(
            vrtDefaultConfig,
          );

          await page.evaluate(() => {
            document.querySelector('bolt-modal').hide();
          });
          await page.waitFor(500);
        }
      },
      timeout,
    );
  });

  modalContent.forEach(async contentChoice => {
    test(
      `${contentChoice.name}<bolt-modal> with band at various viewport sizes`,
      async () => {
        const renderedBand = await render('@bolt-components-band/band.twig', {
          content: contentChoice.content,
          full_bleed: false,
          theme: 'none',
        });

        const { html, ok } = await render('@bolt-components-modal/modal.twig', {
          content: `<bolt-text slot="header">Header slot</bolt-text>
            ${renderedBand.html}
            <bolt-text slot="footer">Footer slot</bolt-text>`,
          width: 'regular',
        });
        expect(ok).toBe(true);
        expect(html).toMatchSnapshot();

        await renderWC('bolt-modal', html, page);

        const screenshots = [];

        for (const item of viewportSizes) {
          const { height, width, size } = item;
          screenshots[size] = [];

          await page.setViewport({ height, width });
          await page.evaluate(() => {
            document.querySelector('bolt-modal').show();
          });
          await page.waitFor(500);

          screenshots[size].modalOpened = await page.screenshot();
          expect(screenshots[size].modalOpened).toMatchImageSnapshot(
            vrtDefaultConfig,
          );

          await page.evaluate(() => {
            document.querySelector('bolt-modal').hide();
          });
          await page.waitFor(500);
        }
      },
      timeout,
    );
  });
});
