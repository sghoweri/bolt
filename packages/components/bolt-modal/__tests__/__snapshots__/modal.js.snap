// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<bolt-modal> Component <bolt-modal> slots 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  <bolt-text slot="header">
    Header slot
  </bolt-text>
  Default slot
  <bolt-text slot="footer">
    Footer slot
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component Long content usage <bolt-modal> at various viewport sizes 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  <bolt-text slot="header">
    Header slot
  </bolt-text>
  <bolt-text subheadline
             font-size="xlarge"
  >
    This is very long content.
  </bolt-text>
  <bolt-image src="/fixtures/1200x2500.jpg"
              alt="Placeholder"
  >
  </bolt-image>
  <bolt-text slot="footer">
    Footer slot
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component Long content usage <bolt-modal> w/o Shadow DOM renders 3`] = `
<bolt-modal
  no-shadow=""
  ready=""
  style=""
  uuid="12345"
  width="regular"
>
  <!---->
  
      
      
  <div
    aria-hidden="true"
    class="c-bolt-modal c-bolt-modal--scroll-container c-bolt-modal--overlay-dark"
    style=""
  >
    
        
    <!---->
    <!---->
    
        
    <focus-trap
      style=""
    >
      
          
      <dialog
        aria-hidden="true"
        aria-labelledby="dialog-title-12345"
        class="c-bolt-modal__content c-bolt-modal__content--width-regular"
        style=""
      >
        
            
        <article
          class="c-bolt-modal__container c-bolt-modal__container--spacing-medium t-bolt-xlight"
          style=""
        >
          
              
          <div
            class="c-bolt-modal__close-button c-bolt-modal__close-button--light"
            style=""
          >
            
                
      
            <bolt-trigger
              autofocus=""
              class="js-close-button-fallback"
              display="block"
              no-outline=""
              style=""
              tabindex="0"
            >
              
        
              <span
                class="c-bolt-modal__close-button__text"
                style=""
              >
                Close this dialog window
              </span>
              
        
              <span
                class="c-bolt-modal__close-button__icon"
                style=""
              />
              
      
            </bolt-trigger>
            
    
              
          </div>
          
              
          <header
            class="c-bolt-modal__container-header"
            style=""
          >
            
                
            <h1
              class="c-bolt-modal__dialog-title"
              id="dialog-title-12345"
              style=""
            >
              
                  Dialog content
                
            </h1>
            
                
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="header"
              style=""
            >
              Header slot
            </bolt-text>
            <!---->
            <!---->
            
              
          </header>
          
              
          <div
            class="c-bolt-modal__container-body"
            style=""
          >
            
                
            <!---->
            <!---->
            
          
            <!---->
            
          
      
            <!---->
            <bolt-text
              class="is-first-child"
              font-size="xlarge"
              style=""
              subheadline=""
            >
              This is very long content.
            </bolt-text>
            <!---->
            
      
            <!---->
            <bolt-image
              alt="Placeholder"
              class="is-last-child"
              src="/fixtures/1200x2500.jpg"
              style=""
            />
            <!---->
            
    
          
            <!---->
            
        
            <!---->
            <!---->
            
              
          </div>
          
              
          
          <footer
            class="c-bolt-modal__container-footer"
            style=""
          >
            
            
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="footer"
              style=""
            >
              Footer slot
            </bolt-text>
            <!---->
            <!---->
            
          
          </footer>
          
        
            
        </article>
        
          
      </dialog>
      
        
    </focus-trap>
    
      
  </div>
  
    
  <!---->
</bolt-modal>
`;

exports[`<bolt-modal> Component Long content usage <bolt-modal> with Shadow DOM renders 3`] = `
<bolt-modal
  ready=""
  style=""
  uuid="12345"
  width="regular"
>
  
          
  <bolt-text
    class="is-first-child is-last-child"
    slot="header"
    style=""
  >
    Header slot
  </bolt-text>
  
          
      
  <bolt-text
    class="is-first-child"
    font-size="xlarge"
    style=""
    subheadline=""
  >
    This is very long content.
  </bolt-text>
  
      
  <bolt-image
    alt="Placeholder"
    class="is-last-child"
    src="/fixtures/1200x2500.jpg"
    style=""
  />
  
    
          
  <bolt-text
    class="is-first-child is-last-child"
    slot="footer"
    style=""
  >
    Footer slot
  </bolt-text>
  
        
</bolt-modal>
`;

exports[`<bolt-modal> Component Long content usage<bolt-modal> with band at various viewport sizes 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  <bolt-text slot="header">
    Header slot
  </bolt-text>
  <bolt-band tag="div"
             size="medium"
             theme="none"
             valign="middle"
  >
    <div class="c-bolt-band c-bolt-band--medium c-bolt-band--valign-middle">
      <div class="c-bolt-band__content">
        <bolt-wrapper size="xxlarge"
                      full="true"
                      bolt-object
        >
          <div class="o-bolt-wrapper o-bolt-wrapper--xxlarge o-bolt-wrapper--full">
            <bolt-text subheadline
                       font-size="xlarge"
            >
              This is very long content.
            </bolt-text>
            <bolt-image src="/fixtures/1200x2500.jpg"
                        alt="Placeholder"
            >
            </bolt-image>
          </div>
        </bolt-wrapper>
      </div>
    </div>
  </bolt-band>
  <bolt-text slot="footer">
    Footer slot
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component Simple usage <bolt-modal> at various viewport sizes 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  <bolt-text slot="header">
    Header slot
  </bolt-text>
  <bolt-text>
    Default slot.
  </bolt-text>
  <bolt-text slot="footer">
    Footer slot
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component Simple usage <bolt-modal> w/o Shadow DOM renders 3`] = `
<bolt-modal
  no-shadow=""
  ready=""
  style=""
  uuid="12345"
  width="regular"
>
  <!---->
  
      
      
  <div
    aria-hidden="true"
    class="c-bolt-modal c-bolt-modal--scroll-container c-bolt-modal--overlay-dark"
    style=""
  >
    
        
    <!---->
    <!---->
    
        
    <focus-trap
      style=""
    >
      
          
      <dialog
        aria-hidden="true"
        aria-labelledby="dialog-title-12345"
        class="c-bolt-modal__content c-bolt-modal__content--width-regular"
        style=""
      >
        
            
        <article
          class="c-bolt-modal__container c-bolt-modal__container--spacing-medium t-bolt-xlight"
          style=""
        >
          
              
          <div
            class="c-bolt-modal__close-button c-bolt-modal__close-button--light"
            style=""
          >
            
                
      
            <bolt-trigger
              autofocus=""
              class="js-close-button-fallback"
              display="block"
              no-outline=""
              style=""
              tabindex="0"
            >
              
        
              <span
                class="c-bolt-modal__close-button__text"
                style=""
              >
                Close this dialog window
              </span>
              
        
              <span
                class="c-bolt-modal__close-button__icon"
                style=""
              />
              
      
            </bolt-trigger>
            
    
              
          </div>
          
              
          <header
            class="c-bolt-modal__container-header"
            style=""
          >
            
                
            <h1
              class="c-bolt-modal__dialog-title"
              id="dialog-title-12345"
              style=""
            >
              
                  Dialog content
                
            </h1>
            
                
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="header"
              style=""
            >
              Header slot
            </bolt-text>
            <!---->
            <!---->
            
              
          </header>
          
              
          <div
            class="c-bolt-modal__container-body"
            style=""
          >
            
                
            <!---->
            <!---->
            
          
            <!---->
            
          
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              style=""
            >
              Default slot.
            </bolt-text>
            <!---->
            
          
            <!---->
            
        
            <!---->
            <!---->
            
              
          </div>
          
              
          
          <footer
            class="c-bolt-modal__container-footer"
            style=""
          >
            
            
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="footer"
              style=""
            >
              Footer slot
            </bolt-text>
            <!---->
            <!---->
            
          
          </footer>
          
        
            
        </article>
        
          
      </dialog>
      
        
    </focus-trap>
    
      
  </div>
  
    
  <!---->
</bolt-modal>
`;

exports[`<bolt-modal> Component Simple usage <bolt-modal> with Shadow DOM renders 3`] = `
<bolt-modal
  ready=""
  style=""
  uuid="12345"
  width="regular"
>
  
          
  <bolt-text
    class="is-first-child is-last-child"
    slot="header"
    style=""
  >
    Header slot
  </bolt-text>
  
          
  <bolt-text
    class="is-first-child is-last-child"
    style=""
  >
    Default slot.
  </bolt-text>
  
          
  <bolt-text
    class="is-first-child is-last-child"
    slot="footer"
    style=""
  >
    Footer slot
  </bolt-text>
  
        
</bolt-modal>
`;

exports[`<bolt-modal> Component Simple usage<bolt-modal> with band at various viewport sizes 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  <bolt-text slot="header">
    Header slot
  </bolt-text>
  <bolt-band tag="div"
             size="medium"
             theme="none"
             valign="middle"
  >
    <div class="c-bolt-band c-bolt-band--medium c-bolt-band--valign-middle">
      <div class="c-bolt-band__content">
        <bolt-wrapper size="xxlarge"
                      full="true"
                      bolt-object
        >
          <div class="o-bolt-wrapper o-bolt-wrapper--xxlarge o-bolt-wrapper--full">
            <bolt-text>
              Default slot.
            </bolt-text>
          </div>
        </bolt-wrapper>
      </div>
    </div>
  </bolt-band>
  <bolt-text slot="footer">
    Footer slot
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component basic usage 1`] = `
<bolt-modal uuid="12345">
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal scroll: container 1`] = `
<bolt-modal scroll="container"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal scroll: overall 1`] = `
<bolt-modal scroll="overall"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: large 1`] = `
<bolt-modal spacing="large"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: medium 1`] = `
<bolt-modal spacing="medium"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: none 1`] = `
<bolt-modal spacing="none"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: small 1`] = `
<bolt-modal spacing="small"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: dark 1`] = `
<bolt-modal theme="dark"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: light 1`] = `
<bolt-modal theme="light"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: none 1`] = `
<bolt-modal theme="none"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: xdark 1`] = `
<bolt-modal theme="xdark"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: xlight 1`] = `
<bolt-modal theme="xlight"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: auto 1`] = `
<bolt-modal width="auto"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: full 1`] = `
<bolt-modal width="full"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: optimal 1`] = `
<bolt-modal width="optimal"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: regular 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;
